### PR TITLE
Regex data sink

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add a data sink interface type which is a receive only endpoint that also acts like an input.  And add regex capability for connecting interfaces.

### CODE REVIEW

File: src/helics/core/Core.hpp [hunk 1]
```diff
@@ -494,6 +494,20 @@ class Core {
     virtual InterfaceHandle getEndpoint(LocalFederateId federateID,
                                         std::string_view name) const = 0;
 
+    /**
+    * Register a data sink which can only receive data from specific targets
+    @param federateID the federate to associate the endpoint with
+    @param name the name of the sink
+    */
```
Comment (nightlark) [inline src/helics/core/Core.hpp:501]:
```suggestion
    @param name the name of the sink
    @return a handle to identify the sink
    */
```

File: src/helics/application_api/MessageFederate.cpp [hunk 2]
```diff
@@ -243,6 +256,19 @@ void MessageFederate::registerMessageInterfacesToml(const std::string& tomlStrin
             Endpoint& epObj =
                 (global) ? registerGlobalEndpoint(key, type) : registerEndpoint(key, type);
 
+            loadOptions(this, ept, epObj);
+        }
+    }
+    if (fileops::isMember(doc, "datasink")) {
+        auto& epts = toml::find(doc, "datasink");
+        if (!epts.is_array()) {
+            throw(helics::InvalidParameter("datasink section in toml file must be an array"));
```
Comment (nightlark) [inline src/helics/application_api/MessageFederate.cpp:265]: `datasinks` for consistency with JSON array name, and matching `endpoints` array name in toml.

```suggestion
    if (fileops::isMember(doc, "datasinks")) {
        auto& epts = toml::find(doc, "datasinks");
        if (!epts.is_array()) {
            throw(helics::InvalidParameter("datasinks section in toml file must be an array"));
```

File: src/helics/core/UnknownHandleManager.cpp [hunk 3]
```diff
@@ -308,39 +308,39 @@ void UnknownHandleManager::clearFilter(const std::string& newFilter)
...
-    }
-    for (auto it = std::begin(unknown_inputs); it != std::end(unknown_inputs);) {
-        if (it->second.first.fed_id == id) {
-            it = unknown_inputs.erase(it);  // previously this was something like m_map.erase(it++);
-        } else {
+    for (auto it = items.begin(); it != items.end();) {
+        if (predicate(*it))
+            it = items.erase(it);
+        else
             ++it;
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/UnknownHandleManager.cpp:319]: I think I tried this and I don't think remove_if works on unordered_map objects, and this code was the recommended pattern until `erase_if` is available.

[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/UnknownHandleManager.cpp:319]: Double checked and remove_if won't work on associative containers, which is why this solution is in place until we start using C++ 20 regularly.

**Review by nightlark — APPROVED** [review]

**Pull request merged**